File: Leetcode/Implement Trie (Prefix Tree).cpp
class TrieNode
{
public:
    TrieNode *next[26];
    bool isleaf;
    TrieNode()
    {
        isleaf = false;
        for (int i = 0; i < 26; i++)
            next[i] = NULL;
    }
};

class Trie
{
    TrieNode *root;

public:
    /** Initialize your data structure here. */
    Trie()
    {
        root = new TrieNode();
    }

    /** Inserts a word into the trie. */
    void insert(string word)
    {
        TrieNode *tmp = root;
        int idx;
        for (int i = 0; i < word.length(); i++)
        {
            idx = word[i] - 'a';
            if (tmp->next[idx] == NULL)
            {
                tmp->next[idx] = new TrieNode();
            }
            tmp = tmp->next[idx];
        }
        tmp->isleaf = true;
    }

    /** Returns if the word is in the trie. */
    bool search(string word)
    {
        TrieNode *tmp = root;
        int idx;
        for (int i = 0; i < word.length(); i++)
        {
            idx = word[i] - 'a';
            tmp = tmp->next[idx];
            if (tmp == NULL)
                return false;
        }
        return tmp->isleaf;
    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix)
    {
        TrieNode *tmp = root;
        int idx;
        for (int i = 0; i < prefix.length(); i++)
        {
            idx = prefix[i] - 'a';
            tmp = tmp->next[idx];
            if (tmp == NULL)
                return false;
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: Leetcode/Implement Trie (Prefix Tree).py
class Trie:

    def __init__(self):
        self.trie = {}

    def insert(self, word: str) -> None:
        t = self.trie
        for w in word:
            if w not in t:
                t[w] = {}
            t = t[w]
        t["#"] = True

    def search(self, word: str) -> bool:
        t = self.trie
        for w in word:
            if w not in t:
                return False
            t = t[w]
        return "#" in t

    def startsWith(self, prefix: str) -> bool:
        t = self.trie
        for w in prefix:
            if w not in t:
                return False
            t = t[w]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
